### DISCUSSION TITLE
What's your overall experience with ALS in your project?

### DISCUSSION THREAD

**Discussion by emrehan98** [Sixze/ALS-Refactored#24] [General]
So I was mingling around some Discord channels, and people saying ALS is nightmare to work because it has too many things and you need to change or make it work within the way of ALS doing things. In some areas, I can agree with that but with my production project, I don't actually have too much trouble. Actually with ALS Refactored having gameplay tags and clean setup with all calculations inside the source, it is pretty a smooth process for me. And I have complicated things like climbing, overlays, swimming and a lot of other things.

What's your experience with ALS? What do you suggest people who use ALS in their project? By the way this question is for everyone. I'd love to hear your experiences and suggestions.

**Comment by Sixze** [comment]
Locomotion systems are usually created individually for each particular game, this is not something that can be easily generalized for all possible use cases.

ALS can be complicated because it is not a Plug and Play solution for all possible types of locomotion, but rather just a very high-quality template, so you need to understand well its internal structure in order to successfully adapt it to your needs. Even though ALS looks amazing, it is quite confusing inside and many things are not done in the best way, which in general I think was the reason why for some people working with ALS was a nightmare.

The main goal of my fork was a complete and high-quality reworking of all the internals of the plugin so that it becomes easier to understand, more flexible and more reliable, especially in multiplayer, and to make it just more pleasant to work with.

Actually, about my experience with ALS, when I was doing a third-person shooter, in fact all I needed was new overlay states for different weapons and animation montages for equipping weapons, shooting and reloading. To do this, it was enough to create a copy of `ABP_Als`, where new overlay states were added and create a character class inherited from `AAlsCharacter`, where I added logic specific to my project.

Also in my project I used the Gameplay Ability System and did all the basic things like shooting or reloading on it and also converted Roll into an ability.

**Reply by emrehan98** [reply]
Thank you for answering. Very inlighting.

I agree with you, I think Longmire did amazing job with ALS and I haven't seen anything close to his work in Unreal Engine community. To me, animation layering was a huge life safer. 

But the problem with the blueprint version of ALS is too many variables, calculations are there, right front of you and when you start changing something, you can easily turn everything into mess. At least that was my experience.

With ALS Refactored my experience is really smooth, only thing I changed within the ALS is LocomotionMode and couple of function accessibility in order to override them within my own cpp class and that was it. I try to keep everything clean and try to stay loyal to base functionality and I don't usually break that rule. Except a few changes for accessing things.

I'll be impelementing much more complex things over the next few weeks, I'm really curious how it will turn out.

**Comment by AchimTuran** [comment]
I use ALS from the community C++ since two years and it was very hard to understand the internals and I still struggle sometimes with the blending curves. I think that's what people find mostly annoying as there is no real documentation at least I don't know it yet.

The C++ implementation is quite good already and I learned a lot from the AnimBP. The initial creator of the Blueprint version of ALS did a very good job in my view and used a lot of features, which I don't know that these exist.

I also adapted the C++ version to be used for side scrollers and made it work with camera volumes to get fixed camera angles working. In my view it is a very high quality system and once you understand to work with it and adapt things if needed you can iterate quite fast.

And to be honest I'm quite amazed how much @Sixze already fixed. Amazing job. The foot locking does so weird things an dI need to test your fixes and migrate them to our fork if they work properly on our side as well.

Furthermore, we have on top of ALS a complete weapon system with physocalöy simulated bullets and an ability system with flying, teleporting, fireballs, double jumps, rolling and wall run. I'm currently also working on a parkour system with motion warping (back ported from UE5, which works very nicely with 4.27). In general I really love ALS and it's amazing to learn from it. Sure the C++ implementation needs a bit of love but it's a good starting point. 😉

**Reply by emrehan98** [reply]
A little bit late but thanks for sharing!
